### ISSUE TITLE
test compute/time requirements for different train tasks

### ISSUE THREAD

**Issue by cmroughan** [Princeton-CDH/htr2hpc#26]
To better automate the creation of slurm batch scripts, we will want a clearer sense of how variations in the training data size and types affects the compute and time requirements for the training job. With that in mind, we will want to run tests evaluating the following:

How does X factor impact time requirements for a training task? What is the optimal number of CPU cores? Run tests addressing these questions for the following factors across both segmentation and transcription training tasks:

- [] File count
- [] File sizes
- [] Complexity of training data
- [] Training from scratch vs fine-tuning

**Comment by cmroughan** [comment]
Consider calculating compile time for producing `.arrow` files for transcription training. Is this always efficient, or only for certain datasets?

**Comment by cmroughan** [comment]
Another round of tests examined what impact different inputs for the `--workers` parameter had on resource usage / train time. Full details are again in the [GSheet](https://docs.google.com/spreadsheets/d/1h5QP_puXG96ioffEV5Nj3Ae5UbKa7JT-30Tzycm8K78) (see where values vary for `worker count`).

Tests were run by changing the parameter for `--workers` in the kraken train command, changing slurm `--cpus-per-task` to match this value, and changing slurm `--mem-per-cpu` so that the total memory across all CPUs stayed approximately the same.

Varying the worker count seems to have had no appreciable impact on transcription training, but does have an impact on segmentation training.

## Transcription

Training from scratch, straight XML:

image count | worker count | mem per CPU | train duration 
---|---|---|---
100|4|4 GB|0:17:09
100|8|2 GB|0:16:27
100|16|1 GB|0:16:27

Training from scratch, binary:

image count | worker count | mem per CPU | train duration
---|---|---|---
100|4|4 GB|0:16:16
100|8|2 GB|0:16:35
100|16|1 GB|0:16:39

## Segmentation

Training "from scratch" (refining on blla.mlmodel):

 image count | worker count | mem per CPU | train duration
---|---|---|---
100|4|12 GB|0:13:51
100|8|6 GB|0:08:49
100|16|3 GB|0:06:01
100|24|2 GB|**OOM killed**\*
100|24|3 GB|0:05:08

\* The fourth entry above never completed an epoch -- it hit an OOM killed error before training began. There seems to be some lower limit on memory per worker required to succeed, depending on the size of the input training data.

Relatedly, for 500 images, training "from scratch" (refining on blla.mlmodel):

image count | worker count | mem per CPU | average epoch
---|---|---|---
500|4|12|724.0 s
500|8|6|491.4 s
500|16|3|**OOM killed**
500|16|4|**OOM killed**

Review of the kraken code might help illuminate what minimums need to be met to avoid the OOM killed error.

**Comment by rlskoeser** [comment]
@cmroughan this is great. Could you also test increasing workers without decreasing memory?

When I was fighting the OOM errors due to the input data problem I did find some comments on github issues about that and vaguely recall seeing 3gb as a minimum for segmentation. I may be able to find again if that would be helpful.

**Comment by cmroughan** [comment]
> Could you also test increasing workers without decreasing memory?

Ran the tests: increasing workers without decreasing memory does not produce tangible benefits, as can be seen in the comparisons below. The code does not end up using the extra resources, which makes it an overly-expensive slurm request and -- if done too often -- can lead to future jobs being ranked lower in the queue as a result.

## Transcription

Training from scratch, straight XML:

image count | worker count | mem per CPU | train duration | CPU usage / requested | percentage
---|---|---|---|---|---
100|4|4 GB|0:17:09|4.2 / 16.0 GB | 26.25%
100|8|4 GB|0:17:13|5.3 / 32.0 GB | 16.56%
100|8|2 GB|0:16:27|5.0 / 16.0 GB | 31.25%

## Segmentation

Training "from scratch" (refining on blla.mlmodel):

image count | worker count | mem per CPU | train duration | CPU usage / requested | percentage
---|---|---|---|---|---
100|4|12 GB|0:13:51|19.7 / 48.0GB | 41.04%
100|8|12 GB|0:09:03|25.1 / 96.0 GB | 26.15%
100|8|6 GB|0:08:49|26.7 / 48.0 GB | 55.63%

**Comment by cmroughan** [comment]
Ran further tests to track how complexity of the training data impacts resource usage, starting with transcription training tasks. For transcription, the line count metric is more relevant than the page count metric (which makes sense, considering the lines *are* the input training data). Line length, here tracked in character counts, also has an impact.

## Tracking line counts

Training from scratch, straight XML, same CPU counts + memory:

image count | line count | avg epoch | full duration 
---|---|---|---
10|524|14.4 s|0:02:24
10|261|6.6 s|0:02:30
5|269|7.4 s|0:02:13
100|5767|161.2|0:15:26
100|2870|83.6 s|0:08:16
47|2845|84.4 s|0:08:17
500|29520|937.8 s|1:22:25
500|14649|464.2 s|0:41:39
241|14563|475.8 s|0:42:37

## Tracking line lengths:

image count | line count | line length | avg epoch
---|---|---|---
500|7538|< 11 chars|113.4 s
500|7537|> 67 chars|361.8 s

In addition to line counts, I'll have to go back and extract the average line lengths across the datasets that have been tested.
